include/swift/SIL/Notifications.h

```diff
@@ -13,9 +13,210 @@
+public:
+  using Handler = void (*)(ModuleDecl *, SILFunction *);
+
+private:
+  Handler handler;
+
+public:
+  FunctionBodyDeserializationNotificationHandler(Handler handler)
+      : handler(handler) {}
+  virtual ~FunctionBodyDeserializationNotificationHandler() {}
+
+  void didDeserializeFunctionBody(ModuleDecl *mod, SILFunction *fn) override {
+    (*handler)(mod, fn);
+  }
+
+  StringRef getName() const override {
+    return "FunctionBodyDeserializationNotificationHandler";
+  }
+};
+
+/// A type that enables us to both pass in custom notification handler types
+/// that use a unique_ptr as well as pass managers and analyses that we want
+/// to pass in a bare pointer for. We use the API
+class DeserializationNotificationHandlerSet
+    : public DeserializationNotificationHandlerBase {
+public:
+  using NotificationUniquePtr =
+    std::unique_ptr<DeserializationNotificationHandler>;
+
+private:
+  /// A list of deserialization callbacks that update the SILModule and other
+  /// parts of SIL as deserialization occurs.
+  ///
+  /// We use 3 here since that is the most that will ever be used today in the
+  /// compiler. If that changed, that number should be changed as well. The
+  /// specific users are:
+  ///
+  /// 1. SILModule's serialization callback.
+  /// 2. SILPassManager notifications.
+  /// 3. Access Enforcement Stripping notification.
+  SmallVector<NotificationUniquePtr, 3> handlerSet;
+
+public:
+  DeserializationNotificationHandlerSet() = default;
+  ~DeserializationNotificationHandlerSet() = default;
+
+  void erase(DeserializationNotificationHandler *handler) {
+    auto iter = find_if(handlerSet, [&](const NotificationUniquePtr &h) {
+      return handler == h.get();
+    });
+    if (iter == handlerSet.end())
+      return;
+    handlerSet.erase(iter);
+  }
+
+  void add(NotificationUniquePtr &&handler) {
+    handlerSet.emplace_back(std::move(handler));
+    sortUnique(handlerSet);
+  }
+
+  static DeserializationNotificationHandler *
+  getUnderlyingHandler(const NotificationUniquePtr &h) {
+    return h.get();
+  }
+
+  using iterator = llvm::mapped_iterator<
+      decltype(handlerSet)::const_iterator,
+      decltype(&DeserializationNotificationHandlerSet::getUnderlyingHandler)>;
+
+  iterator begin() const {
+    auto *fptr = &DeserializationNotificationHandlerSet::getUnderlyingHandler;
+    return llvm::map_iterator(handlerSet.begin(), fptr);
+  }
+
+  iterator end() const {
+    auto *fptr = &DeserializationNotificationHandlerSet::getUnderlyingHandler;
+    return llvm::map_iterator(handlerSet.end(), fptr);
+  }
+
+  using range = llvm::iterator_range<iterator>;
+  auto getRange() const -> range { return llvm::make_range(begin(), end()); }
+
+  // DeserializationNotificationHandler implementation via chaining to the
+  // handlers we contain.
+#ifdef DNS_CHAIN_METHOD
+#error "DNS_CHAIN_METHOD is defined?!"
+#endif
+#define DNS_CHAIN_METHOD(Name, FirstTy, SecondTy)                              \
```

Comment:
I'm not at all seeing a net benefit of macros here.

Reply:
Gone!